File: sdk/flutter_analytics/android/src/main/kotlin/com/myampmix/analytics/MyampmixAnalyticsPlugin.kt
package com.myampmix.analytics

import android.content.Context
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.BillingClientStateListener
import com.android.billingclient.api.BillingResult
import com.android.billingclient.api.Purchase
import com.android.billingclient.api.PurchasesUpdatedListener
import com.android.billingclient.api.QueryProductDetailsParams
import com.android.billingclient.api.QueryPurchasesParams
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel

/**
 * MyAmpMix native store-purchase autocapture (Android half).
 *
 * Wraps a Google Play [BillingClient] purely as an OBSERVER of the app's own
 * in-app-purchase transactions — the same role RevenueCat/Adjust play when
 * they hook Play Billing. This plugin never calls `launchBillingFlow`,
 * never acknowledges a purchase, and never consumes one: fulfilment stays
 * the responsibility of whatever code in the host app actually granted the
 * purchase. Forwarded payloads are consumed by
 * `lib/src/autocapture/purchase_autocapture.dart` over the
 * `myampmix_analytics/purchases` [EventChannel] and re-emitted through the
 * Dart facade as the reserved `$in_app_purchase` event.
 *
 * IMPORTANT ANDROID CAVEAT (unlike iOS, where `SKPaymentQueue` is a single
 * process-wide queue any observer can attach to): Play Billing's
 * [PurchasesUpdatedListener] only fires for purchases launched through THIS
 * SAME [BillingClient] instance. If the host app purchases through its own,
 * separate `BillingClient` (e.g. RevenueCat's), this listener never
 * observes that purchase flow directly. To still surface those, this
 * plugin also calls `queryPurchasesAsync` once its client connects, which
 * returns every INAPP purchase owned by the signed-in Play account for this
 * app package — regardless of which client/session originally launched it
 * — mirroring how StoreKit replays existing transactions to a new
 * observer. Either way, this is still only ever the APP'S OWN
 * transactions: there is no API for observing another app's or another
 * user's purchases (that is all a sandboxed app can ever see).
 *
 * Kept deliberately defensive: every entry point is wrapped so a Play
 * Billing failure (missing Play Services, library mismatch, disconnects,
 * malformed callback data, ...) never crashes the host app.
 */
class MyampmixAnalyticsPlugin :
    FlutterPlugin,
    PurchasesUpdatedListener {

    private var eventChannel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null
    private var billingClient: BillingClient? = null

    /** De-dupes a purchase seen both via the live listener and the replay query. */
    private val seenTransactionIds = mutableSetOf<String>()

    /** Buffers payloads observed before Dart attaches a stream listener. */
    private val pendingPayloads = mutableListOf<Map<String, Any?>>()

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        val channel = EventChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setStreamHandler(
            object : EventChannel.StreamHandler {
                override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
                    eventSink = events
                    if (pendingPayloads.isNotEmpty()) {
                        pendingPayloads.forEach { events.success(it) }
                        pendingPayloads.clear()
                    }
                }

                override fun onCancel(arguments: Any?) {
                    eventSink = null
                }
            },
        )
        eventChannel = channel
        startBillingClient(binding.applicationContext)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        eventChannel?.setStreamHandler(null)
        eventChannel = null
        eventSink = null
        try {
            billingClient?.endConnection()
        } catch (_: Throwable) {
            // Never crash the host on teardown.
        }
        billingClient = null
    }

    private fun startBillingClient(context: Context) {
        try {
            val client = BillingClient.newBuilder(context)
                .setListener(this)
                .enablePendingPurchases()
                .build()
            billingClient = client
            client.startConnection(
                object : BillingClientStateListener {
                    override fun onBillingSetupFinished(billingResult: BillingResult) {
                        if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                            replayExistingPurchases(client)
                        }
                    }

                    override fun onBillingServiceDisconnected() {
                        // Never crash; a future purchase update (if the
                        // client reconnects) simply resumes forwarding.
                    }
                },
            )
        } catch (_: Throwable) {
            // Defensive: Play Billing may be unavailable at runtime (no
            // Play Services, library mismatch, ...). Never crash the host.
        }
    }

    /**
     * Replays purchases that already exist for this account/app so a
     * purchase made through a DIFFERENT billing client (see class doc)
     * still gets picked up, much like StoreKit replays its queue to a
     * freshly attached observer.
     */
    private fun replayExistingPurchases(client: BillingClient) {
        try {
            val params = QueryPurchasesParams.newBuilder()
                .setProductType(BillingClient.ProductType.INAPP)
                .build()
            client.queryPurchasesAsync(params) { billingResult, purchases ->
                if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                    purchases.forEach { handlePurchase(client, it) }
                }
            }
        } catch (_: Throwable) {
            // Best-effort only.
        }
    }

    override fun onPurchasesUpdated(billingResult: BillingResult, purchases: MutableList<Purchase>?) {
        try {
            if (billingResult.responseCode != BillingClient.BillingResponseCode.OK) return
            val client = billingClient ?: return
            purchases?.forEach { handlePurchase(client, it) }
        } catch (_: Throwable) {
            // Never crash the host from a billing callback.
        }
    }

    private fun handlePurchase(client: BillingClient, purchase: Purchase) {
        try {
            if (purchase.purchaseState != Purchase.PurchaseState.PURCHASED) return
            val transactionId = purchase.orderId ?: purchase.purchaseToken
            if (transactionId.isNullOrEmpty()) return
            if (!seenTransactionIds.add(transactionId)) return // already forwarded

            fetchProductPricing(client, purchase.products) { priceMicros, currency ->
                if (purchase.products.isEmpty()) return@fetchProductPricing
                purchase.products.forEach { productId ->
                    emit(
                        mapOf(
                            "productId" to productId,
                            "price" to priceMicros?.let { it / 1_000_000.0 },
                            "currency" to currency,
                            "quantity" to purchase.quantity,
                            "transactionId" to transactionId,
                            "store" to "play_store",
                        ),
                    )
                }
            }
        } catch (_: Throwable) {
            // Never crash the host from a billing callback.
        }
    }

    /**
     * Best-effort price/currency lookup for the purchase's first product
     * via `queryProductDetailsAsync`. Play Billing purchases do not carry
     * price themselves — only `ProductDetails` does. Failure (or an app
     * with no matching product details, e.g. mismatched sandbox config)
     * degrades to `null`/`null`, never a crash.
     */
    private fun fetchProductPricing(
        client: BillingClient,
        productIds: List<String>,
        onResult: (priceMicros: Long?, currency: String?) -> Unit,
    ) {
        val firstProductId = productIds.firstOrNull()
        if (firstProductId == null) {
            onResult(null, null)
            return
        }
        try {
            val product = QueryProductDetailsParams.Product.newBuilder()
                .setProductId(firstProductId)
                .setProductType(BillingClient.ProductType.INAPP)
                .build()
            val params = QueryProductDetailsParams.newBuilder()
                .setProductList(listOf(product))
                .build()
            client.queryProductDetailsAsync(params) { billingResult, productDetailsList ->
                val offer = productDetailsList.firstOrNull()?.oneTimePurchaseOfferDetails
                if (billingResult.responseCode == BillingClient.BillingResponseCode.OK && offer != null) {
                    onResult(offer.priceAmountMicros, offer.priceCurrencyCode)
                } else {
                    onResult(null, null)
                }
            }
        } catch (_: Throwable) {
            onResult(null, null)
        }
    }

    private fun emit(payload: Map<String, Any?>) {
        val sink = eventSink
        if (sink != null) {
            sink.success(payload)
        } else {
            pendingPayloads.add(payload)
        }
    }

    private companion object {
        const val CHANNEL_NAME = "myampmix_analytics/purchases"
    }
}

File: sdk/flutter_analytics/ios/Classes/MyampmixAnalyticsPlugin.swift
import Flutter
import StoreKit
import UIKit

/// MyAmpMix native store-purchase autocapture (iOS half).
///
/// Registers an `SKPaymentTransactionObserver` on the shared, process-wide
/// `SKPaymentQueue` and forwards `.purchased`/`.restored` transactions to
/// Dart over the `myampmix_analytics/purchases` `EventChannel`. Consumed by
/// `lib/src/autocapture/purchase_autocapture.dart`, which re-emits them
/// through the Dart facade as the reserved `$in_app_purchase` event.
///
/// IMPORTANT — this is a PASSIVE observer, exactly like RevenueCat/Adjust's
/// StoreKit hooks: it never calls `finishTransaction`. Finishing/acking a
/// transaction remains the responsibility of whatever code in the host app
/// actually fulfils the purchase (the app's own StoreKit integration,
/// RevenueCat, ...). We only look; we never touch.
///
/// A sandboxed plugin can only ever see transactions that pass through this
/// process's `SKPaymentQueue` — i.e. purchases made by THIS app install.
/// There is no API for observing another app's or another user's
/// purchases; "native" here means "the app's own StoreKit transactions",
/// not store-wide telemetry.
///
/// Kept deliberately defensive: every callback is wrapped so a StoreKit
/// failure or unexpected transaction shape never crashes the host app.
public class MyampmixAnalyticsPlugin: NSObject, FlutterPlugin, SKPaymentTransactionObserver, SKProductsRequestDelegate, FlutterStreamHandler {
  private static let channelName = "myampmix_analytics/purchases"

  private var eventSink: FlutterEventSink?

  /// Transactions observed before Dart attaches a stream listener (e.g. a
  /// restored transaction replayed at cold start, before `runApp`
  /// completes) are buffered here and flushed as soon as `onListen` runs,
  /// so nothing is silently dropped.
  private var pendingPayloads: [[String: Any?]] = []

  /// Best-effort product price/currency cache, keyed by productIdentifier.
  /// Populated asynchronously via `SKProductsRequest`; a transaction is
  /// forwarded immediately regardless of whether this cache is warm yet —
  /// price/currency are simply `nil` until/unless a lookup completes.
  private var productCache: [String: (price: Double, currency: String?)] = [:]
  private var pendingProductRequests: Set<String> = []
  private var productIdsByRequest: [ObjectIdentifier: [String]] = [:]

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = MyampmixAnalyticsPlugin()
    let eventChannel = FlutterEventChannel(
      name: channelName,
      binaryMessenger: registrar.messenger()
    )
    eventChannel.setStreamHandler(instance)
    SKPaymentQueue.default().add(instance)
    // Keeps the plugin instance alive for the lifetime of the engine so the
    // SKPaymentQueue observer registration above is never deallocated.
    registrar.publish(instance)
  }

  deinit {
    SKPaymentQueue.default().remove(self)
  }

  // MARK: - FlutterStreamHandler

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    if !pendingPayloads.isEmpty {
      pendingPayloads.forEach { events($0) }
      pendingPayloads.removeAll()
    }
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  // MARK: - SKPaymentTransactionObserver

  public func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {
    for transaction in transactions {
      switch transaction.transactionState {
      case .purchased, .restored:
        forward(transaction)
        maybeFetchProduct(for: transaction.payment.productIdentifier)
      default:
        break
      }
    }
  }

  private func forward(_ transaction: SKPaymentTransaction) {
    let productId = transaction.payment.productIdentifier
    let transactionId = transaction.transactionIdentifier ?? transaction.original?.transactionIdentifier
    guard !productId.isEmpty, let transactionId = transactionId, !transactionId.isEmpty else {
      // Never forward a payload the Dart side cannot map onto the §4
      // contract's required properties.
      return
    }
    let cached = productCache[productId]
    let payload: [String: Any?] = [
      "productId": productId,
      "price": cached?.price,
      "currency": cached?.currency,
      "quantity": transaction.payment.quantity,
      "transactionId": transactionId,
      "store": "app_store",
    ]
    if let sink = eventSink {
      sink(payload)
    } else {
      pendingPayloads.append(payload)
    }
  }

  private func maybeFetchProduct(for productId: String) {
    guard productCache[productId] == nil, !pendingProductRequests.contains(productId) else { return }
    pendingProductRequests.insert(productId)
    let request = SKProductsRequest(productIdentifiers: [productId])
    request.delegate = self
    productIdsByRequest[ObjectIdentifier(request)] = [productId]
    request.start()
  }

  // MARK: - SKProductsRequestDelegate

  public func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse) {
    let requestedIds = productIdsByRequest.removeValue(forKey: ObjectIdentifier(request)) ?? []
    for product in response.products {
      pendingProductRequests.remove(product.productIdentifier)
      productCache[product.productIdentifier] = (
        price: product.price.doubleValue,
        currency: product.priceLocale.currencyCode
      )
    }
    for id in requestedIds where productCache[id] == nil {
      pendingProductRequests.remove(id)
    }
  }

  public func request(_ request: SKRequest, didFailWithError error: Error) {
    // Best-effort only: leave price/currency nil for these products next
    // time they're forwarded. Never crash on a products-request failure.
    guard let productsRequest = request as? SKProductsRequest else { return }
    let requestedIds = productIdsByRequest.removeValue(forKey: ObjectIdentifier(productsRequest)) ?? []
    requestedIds.forEach { pendingProductRequests.remove($0) }
  }
}
